datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -617,11 +862,78 @@ where
                         }
                         block_ptrs
                     }
-                    _ => unreachable!(),
                 }),
         )
     }
 
+    fn blocks_with_calls(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+        call_filter: EthereumCallFilter,
+    ) -> Box<Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send> {
+        let eth = self.clone();
+
+        let addresses: Vec<H160> = call_filter
+            .contract_addresses_function_signatures
+            .iter()
+            .map(|(addr, _fsigs)| *addr)
+            .collect::<HashSet<H160>>()
+            .into_iter()
+            .collect::<Vec<H160>>();
+        let blocks = eth
+            .trace_stream(&logger, from, to, addresses)
+            .collect()
+            .map(move |trace_chunks| {
+                match trace_chunks.len() {
+                    0 => vec![],
+                    _ => {
+                        trace_chunks
+                            .iter()
+                            .flatten()
+                            .filter(|trace| {
+                                let is_call = match trace.action {
+                                    Action::Call(_) => true,
+                                    _ => false,
+                                };
+                                // Remove traces that are not for a call, do not have a result, or
+                                // are for a transaction which errored.
+                                if !is_call || trace.result.is_none() || trace.error.is_some() {
+                                    return false;
+                                }
+                                true
+                            })
+                            .map(EthereumCall::from)
+                            .filter(|call| {
+                                // `trace_filter` can only filter by calls `to` an address and
+                                // a block range. Since subgraphs are subscribing to calls
+                                // for a specific contract function an additional filter needs
+                                // to be applied
+                                call_filter.matches(&call)
+                            })
+                            .collect()
+                    }
+                }
+            })
+            .map(|calls| {
+                let mut block_ptrs = vec![];
+                for call in calls.iter() {
+                    let hash = call.block_hash;
+                    let number = call.block_number;
+                    let block_ptr = EthereumBlockPointer::from((hash, number));
```

Comment:
I'd pass `call.block_hash` and `call.block_number` to `EthereumBlockPointer::from` directly.